test/Marketplace.js

```diff
@@ -70,199 +89,276 @@ contract('Marketplace', function([_, owner, seller, buyer]) {
     })
 
-    // Event emitted
-    logs.length.should.be.equal(1)
-    checkAuctionCreatedLog(logs[0], assetId, seller, itemPrice, endTime)
-
-    // Check data
-    let s = await market.auctionByAssetId(assetId)
-    s[1].should.be.equal(seller)
-    s[2].should.be.bignumber.equal(itemPrice)
-    s[3].should.be.bignumber.equal(endTime)
-  })
+    it('should update an order', async function() {
+      let newPrice = web3.toWei(2.0, 'ether')
+      let newEndTime = endTime + duration.minutes(5)
+
+      const { logs } = await market.createOrder(
+        assetId,
+        erc721.address,
+        newPrice,
+        newEndTime,
+        { from: seller }
+      )
+
+      // Event emitted
+      logs.length.should.be.equal(1)
+      checkOrderCreatedLog(
+        logs[0],
+        assetId,
+        seller,
+        erc721.address,
+        newPrice,
+        newEndTime
+      )
+
+      // Check data
+      let s = await market.orderByAssetId.call(erc721.address, assetId)
+      s[1].should.be.equal(seller)
+      s[2].should.be.equal(erc721.address)
+      s[3].should.be.bignumber.equal(newPrice)
+      s[4].should.be.bignumber.equal(newEndTime)
+    })
 
-  it('should update an order', async function() {
-    let newPrice = web3.toWei(2.0, 'ether')
-    let newEndTime = endTime + duration.minutes(5)
+    it('should fail to create an order :: (contract not approved)', async function() {
+      const newAssetId = 123123123
+      await erc721.mint(otherAddress, newAssetId)
 
-    const { logs } = await market.createOrder(assetId, newPrice, newEndTime, {
-      from: seller
+      await market
+        .createOrder(assetId, erc721.address, itemPrice, endTime, {
+          from: otherAddress
+        })
+        .should.be.rejectedWith(EVMRevert)
     })
 
-    // Event emitted
-    logs.length.should.be.equal(1)
-    checkAuctionCreatedLog(logs[0], assetId, seller, newPrice, newEndTime)
+    it('should fail to create an order :: (address not the owner of asset)', async function() {
+      await market
+        .createOrder(assetId, erc721.address, itemPrice, endTime, {
+          from: otherAddress
+        })
+        .should.be.rejectedWith(EVMRevert)
+    })
 
-    // Check data
-    let s = await market.auctionByAssetId(assetId)
-    s[1].should.be.equal(seller)
-    s[2].should.be.bignumber.equal(newPrice)
-    s[3].should.be.bignumber.equal(newEndTime)
+    it('should fail to create an order :: (not an ERC721 contract)', async function() {
+      await market
+        .createOrder(assetId, erc20.address, itemPrice, endTime, {
+          from: seller
+        })
+        .should.be.rejectedWith(EVMRevert)
+    })
   })
 
-  // Cancel
-
-  it('should cancel a created order', async function() {
-    let itemPrice = web3.toWei(1.0, 'ether')
-
-    await market.createOrder(assetId, itemPrice, endTime, {
-      from: seller
+  describe('Cancel', function() {
+    it('should cancel a created order', async function() {
```

Comment:
Maybe we can add a test to see if the contract owner can cancel the order as it's one of the conditions in `require(order.seller == msg.sender || msg.sender == owner, "Unauthorized user");`

Reply:
Maybe we could write a test to Double Cancel an order. This will test that we are correctly removing the Order from the mapping.